path_login.go

```diff
@@ -88,129 +86,33 @@ func (b *jwtAuthBackend) pathLogin(ctx context.Context, req *logical.Request, d
-		}
-
-		if *claims.Expiry == 0 {
-			latestStart := *claims.IssuedAt
-			if *claims.NotBefore > *claims.IssuedAt {
-				latestStart = *claims.NotBefore
-			}
-			leeway := role.ExpirationLeeway.Seconds()
-			if role.ExpirationLeeway.Seconds() < 0 {
-				leeway = 0
-			} else if role.ExpirationLeeway.Seconds() == 0 {
-				leeway = claimDefaultLeeway
-			}
-			*claims.Expiry = jwt.NumericDate(int64(latestStart) + int64(leeway))
-		}
-
-		if *claims.NotBefore == 0 {
-			if *claims.IssuedAt != 0 {
-				*claims.NotBefore = *claims.IssuedAt
-			} else {
-				leeway := role.NotBeforeLeeway.Seconds()
-				if role.NotBeforeLeeway.Seconds() < 0 {
-					leeway = 0
-				} else if role.NotBeforeLeeway.Seconds() == 0 {
-					leeway = claimDefaultLeeway
-				}
-				*claims.NotBefore = jwt.NumericDate(int64(*claims.Expiry) - int64(leeway))
-			}
-		}
-
-		if len(claims.Audience) > 0 && len(role.BoundAudiences) == 0 {
-			return logical.ErrorResponse("audience claim found in JWT but no audiences bound to the role"), nil
-		}
-
-		expected := jwt.Expected{
-			Issuer:  config.BoundIssuer,
-			Subject: role.BoundSubject,
-			Time:    time.Now(),
-		}
-
-		cksLeeway := role.ClockSkewLeeway
-		if role.ClockSkewLeeway.Seconds() < 0 {
-			cksLeeway = 0
-		} else if role.ClockSkewLeeway.Seconds() == 0 {
-			cksLeeway = jwt.DefaultLeeway
-		}
-
-		if err := claims.ValidateWithLeeway(expected, cksLeeway); err != nil {
-			return logical.ErrorResponse(errwrap.Wrapf("error validating claims: {{err}}", err).Error()), nil
-		}
+	// Get the JWT validator based on the configured auth type
+	validator, err := b.getJWTValidator(config)
+	if err != nil {
+		return logical.ErrorResponse(errwrap.Wrapf("error configuring token validator: {{err}}", err).Error()), nil
+	}
 
-		if err := validateAudience(role.BoundAudiences, claims.Audience, true); err != nil {
-			return logical.ErrorResponse(errwrap.Wrapf("error validating claims: {{err}}", err).Error()), nil
-		}
+	// Set expected claims values to assert on the JWT
+	expected := jwt.Expected{
+		Issuer:            config.BoundIssuer,
+		Subject:           role.BoundSubject,
+		Audiences:         role.BoundAudiences,
+		SigningAlgorithms: toAlg(config.JWTSupportedAlgs),
+		NotBeforeLeeway:   role.NotBeforeLeeway,
+		ExpirationLeeway:  role.ExpirationLeeway,
+		ClockSkewLeeway:   role.ClockSkewLeeway,
+	}
 
-	case configType == OIDCDiscovery:
-		allClaims, err = b.verifyOIDCToken(ctx, config, role, token)
-		if err != nil {
-			return logical.ErrorResponse(err.Error()), nil
-		}
+	// Validate the JWT by verifying it's signature and asserting expected claims values
+	allClaims, err := validator.Validate(ctx, token, expected)
+	if err != nil {
+		return logical.ErrorResponse(errwrap.Wrapf("error validating token: {{err}}", err).Error()), nil
+	}
 
-	default:
-		return nil, errors.New("unhandled case during login")
+	// If the role has no bound audiences, then return an error if any audience in the JWT claim
```

Comment:
The comment here seems to be incomplete.

Reply:
I've updated the comment in 1ff98b6. Let me know if that makes more sense!